### PR TITLE
Simplify RecordBatchReader to use Project.next()

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

File: cpp/src/lance/io/exec/project.cc [hunk 1]
```diff
@@ -18,6 +18,7 @@
 #include <arrow/result.h>
 #include <fmt/format.h>
 
+#include "lance/arrow/file_lance_ext.h"
```
Comment (eddyxu) [inline cpp/src/lance/io/exec/project.cc:21]: Forgot to push these changes from the last PR.

File: cpp/src/lance/io/exec/project.cc [hunk 2]
```diff
@@ -31,16 +32,23 @@ Project::Project(std::unique_ptr<ExecNode> child) : child_(std::move(child)) {}
...
   auto projected_arrow_schema = scan_options->projected_schema;
   if (projected_arrow_schema->num_fields() == 0) {
     projected_arrow_schema = scan_options->dataset_schema;
   }
   ARROW_ASSIGN_OR_RAISE(auto projected_schema, schema.Project(*projected_arrow_schema));
 
+  std::optional<int64_t> limit;
+  int64_t offset;
+  if (scan_options->fragment_scan_options &&
+      scan_options->fragment_scan_options->type_name() == "lance") {
```
Comment (changhiskhan) [inline cpp/src/lance/io/exec/project.cc:46]: Let's extract the string check into a util function?

**Review by changhiskhan — APPROVED** [review]
one nit

File: cpp/src/lance/io/exec/project.cc (hunk 2 above)
Comment (eddyxu) [reply cpp/src/lance/io/exec/project.cc]: done

**Pull request merged**